I was about to merge the PR but then I noticed that the Windows build fails:

```
=================================== 31/31 ====================================
test:         nvme-cli:nvme-cli - command-metadata-schema
start time:   22:07:39
duration:     0.21s
result:       exit status 1
command:      MALLOC_PERTURB_=36 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 PATH=D:/a/nvme-cli/nvme-cli/.build-ci/libnvme/src;D:/a/nvme-cli/nvme-cli/.build-ci/;D:\a\_temp\msys64\ucrt64\bin;D:\a\_temp\msys64\usr\local\bin;D:\a\_temp\msys64\usr\bin;D:\a\_temp\msys64\usr\bin;C:\Windows\System32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;D:\a\_temp\msys64\usr\bin\site_perl;D:\a\_temp\msys64\usr\bin\vendor_perl;D:\a\_temp\msys64\usr\bin\core_perl D:\a\_temp\msys64\ucrt64\bin/python3.EXE D:/a/nvme-cli/nvme-cli/.build-ci/../unit-py/test_command_metadata_schema.py ./nvme.exe D:/a/nvme-cli/nvme-cli/plugins/utils/command-metadata-schema.json
----------------------------------- stderr -----------------------------------
E
======================================================================
ERROR: setUpClass (__main__.TestCommandMetadataSchema)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:/a/nvme-cli/nvme-cli/.build-ci/../unit-py/test_command_metadata_schema.py", line 133, in setUpClass
    raise AssertionError(
        "dump-command-metadata produced no output (built with json-c)")
AssertionError: dump-command-metadata produced no output (built with json-c)

----------------------------------------------------------------------
```
From the build logs, it looks like there is json-c support enabled. So not clear what's going on here.

BTW, I think you can squash all three patches together. I would have done that anyway :)

--- a/plugins/utils/command-metadata.c
+++ b/plugins/utils/command-metadata.c
@@ -21,10 +21,14 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <getopt.h>
+#include <signal.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 #include "command-metadata.h"
 #include "common.h"
@@ -63,6 +67,139 @@ static struct command_metadata_command *command_metadata_cur_command;
  * for the parser-unwind sentinel; checked after each command fn returns.
  */
 static int command_metadata_capture_error;
+
+/*
+ * capture_saved_stderr_fd holds a dup of stderr from before the NUL redirect,
+ * so crash messages still reach the user.
+ */
+static const char *capture_current_command;
+static int capture_saved_stderr_fd = -1;
+
+/*
+ * (void)write() does not suppress GCC's warn_unused_result under
+ * _FORTIFY_SOURCE; assigning to a discarded variable does.
+ */
+static void write_raw(int fd, const char *buf, size_t len)
+{
+	ssize_t ret = write(fd, buf, len);
+	(void)ret;
+}
+
+/* strlen() is not async-signal-safe; measure inline for the crash handler. */
+static void write_str(int fd, const char *s)
+{
+	size_t len = 0;
+
+	if (!s)
+		return;
+
+	while (s[len])
+		len++;
+
+	write_raw(fd, s, len);
+}
+
+static void write_uint(int fd, unsigned int n)
+{
+	char buf[10];	/* enough for any 32-bit unsigned int */
+	int i = sizeof(buf);
+
+	do {
+		buf[--i] = '0' + (n % 10);
+		n /= 10;
+	} while (n && i);
+
+	write_raw(fd, buf + i, sizeof(buf) - i);
+}
+
+static void capture_crash_handler(int sig)
+{
+	int fd = capture_saved_stderr_fd >= 0 ? capture_saved_stderr_fd : STDERR_FILENO;
+
+	write_str(fd, "dump-command-metadata: fatal: '");
+	write_str(fd, capture_current_command ? capture_current_command : "(unknown)");
+	write_str(fd, "' crashed during option capture (signal ");
+	write_uint(fd, sig);
+	write_str(fd, ")\n");
+
+	signal(sig, SIG_DFL);
+	raise(sig);
+}
+
+/*
+ * On Windows, OS-level exceptions (stack overflow, access violations) bypass
+ * the C signal() mechanism entirely and terminate the process silently.
+ * SetUnhandledExceptionFilter intercepts these so we can print the same
+ * diagnostic before dying.
+ */
+#ifdef _WIN32
+static void write_hex(int fd, unsigned int n)
+{
+	char buf[8];
+	int i = sizeof(buf);
+
+	do {
+		buf[--i] = "0123456789abcdef"[n & 0xf];
+		n >>= 4;
+	} while (n && i);
+
+	write_raw(fd, buf + i, sizeof(buf) - i);
+}
+
+static LONG WINAPI capture_exception_filter(EXCEPTION_POINTERS *ep)
+{
+	int fd = capture_saved_stderr_fd >= 0 ? capture_saved_stderr_fd : STDERR_FILENO;
+
+	write_str(fd, "dump-command-metadata: fatal: '");
+	write_str(fd, capture_current_command ? capture_current_command : "(unknown)");
+	write_str(fd, "' crashed during option capture (exception 0x");
+	write_hex(fd, ep->ExceptionRecord->ExceptionCode);
+	write_str(fd, ")\n");
+
+	return EXCEPTION_CONTINUE_SEARCH;
+}
+
+static LPTOP_LEVEL_EXCEPTION_FILTER capture_prev_filter;
+#endif
+
+static void (*capture_prev_sigsegv)(int);
+static void (*capture_prev_sigabrt)(int);
+#ifdef SIGBUS
+static void (*capture_prev_sigbus)(int);
+#endif
+
+/* Normalize SIG_ERR to SIG_DFL so restoration is always a valid disposition. */
+static void (*set_crash_handler(int sig))(int)
+{
+	void (*prev)(int) = signal(sig, capture_crash_handler);
+
+	return prev == SIG_ERR ? SIG_DFL : prev;
+}
+
+static void install_crash_handlers(void)
+{
+	capture_prev_sigsegv = set_crash_handler(SIGSEGV);
+	capture_prev_sigabrt = set_crash_handler(SIGABRT);
+#ifdef SIGBUS
+	capture_prev_sigbus = set_crash_handler(SIGBUS);
+#endif
+#ifdef _WIN32
+	capture_prev_filter = SetUnhandledExceptionFilter(capture_exception_filter);
+#endif
+}
+
+static void remove_crash_handlers(void)
+{
+	signal(SIGSEGV, capture_prev_sigsegv);
+	signal(SIGABRT, capture_prev_sigabrt);
+#ifdef SIGBUS
+	signal(SIGBUS, capture_prev_sigbus);
+#endif
+#ifdef _WIN32
+	SetUnhandledExceptionFilter(capture_prev_filter);
+	capture_prev_filter = NULL;
+#endif
+}
 
 /* ------------------------------------------------------------------ */
 /* Pass 1: capture                                                    */
@@ -215,7 +352,9 @@ static int capture_command(struct command_metadata_command *mc, struct command *
 
 	command_metadata_capture_error = 0;
 	command_metadata_cur_command = mc;
+	capture_current_command = cmd->name;
 	(void)cmd->fn(2, argv, cmd, plugin);
+	capture_current_command = NULL;
 	command_metadata_cur_command = NULL;
 
 	/*
@@ -281,12 +420,14 @@ static struct command_metadata_program *build_model(struct program *prog)
 	if (devnull >= 0) {
 		saved_stdout = dup(STDOUT_FILENO);
 		saved_stderr = dup(STDERR_FILENO);
+		capture_saved_stderr_fd = saved_stderr;
 		if (saved_stdout >= 0)
 			dup2(devnull, STDOUT_FILENO);
 		if (saved_stderr >= 0)
 			dup2(devnull, STDERR_FILENO);
 	}
 
+	install_crash_handlers();
 	argconfig_set_parse_hook(metadata_capture_hook);
 
 	for (pi = 0, plugin = prog->extensions; plugin; plugin = plugin->next, pi++) {
@@ -316,6 +457,8 @@ static struct command_metadata_program *build_model(struct program *prog)
 	}
 
 	argconfig_set_parse_hook(NULL);
+	remove_crash_handlers();
+	capture_saved_stderr_fd = -1;
 
 	fflush(stdout);
 	fflush(stderr);

--- a/unit-py/test_command_metadata_schema.py
+++ b/unit-py/test_command_metadata_schema.py
@@ -63,8 +63,8 @@ def json_c_available():
 
 
 def dump_metadata():
-    """Run `nvme utils dump-command-metadata` and return parsed JSON, or None if
-    it produced no usable output."""
+    """Run `nvme utils dump-command-metadata` and return (parsed JSON, stderr).
+    Returns (None, stderr) if it produced no usable output."""
     proc = subprocess.run(
         [NVME_BIN, "utils", "dump-command-metadata"],
         stdout=subprocess.PIPE,
@@ -73,8 +73,8 @@ def dump_metadata():
     )
     text = proc.stdout.strip()
     if proc.returncode != 0 or not text:
-        return None
-    return json.loads(text)
+        return None, proc.stderr
+    return json.loads(text), proc.stderr
 
 
 def iter_commands(data):
@@ -128,10 +128,12 @@ class TestCommandMetadataSchema(unittest.TestCase):
     def setUpClass(cls):
         if not json_c_available():
             raise unittest.SkipTest("nvme built without json-c")
-        cls.data = dump_metadata()
+        cls.data, stderr = dump_metadata()
         if cls.data is None:
-            raise AssertionError(
-                "dump-command-metadata produced no output (built with json-c)")
+            msg = "dump-command-metadata produced no output (built with json-c)"
+            if stderr:
+                msg += "\nstderr: " + stderr.rstrip()
+            raise AssertionError(msg)
         with open(SCHEMA_PATH) as f:
             cls.schema = json.load(f)
 

